telegram-translation-shortcuts.user.js

```diff
@@ -66,20 +67,78 @@ function scrollItems (down) {
   }
 }
 
+/**
+ * A div (key-add-suggestion-wrap) contains the input-form to Add translation with Submit and Cancel buttons
+ * This function will check if that div is collapsed or not. If it is collapsed,
+ * it's className will change to 'key-add-suggestion-wrap collapsed'
+ *
+ * @function inputWrapCollapsed() -- Checks if element by class 'key-add-suggestion-wrap' is exactly the same. If not the same, then it is collapsed
+ * @returns True if collapsed
+ * @returns False if NOT collapsed
+ */
+function inputWrapCollapsed () {
+  if (document.getElementsByClassName('key-add-suggestion-wrap')[0].className === 'key-add-suggestion-wrap') { // checking if it is NOT collapsed
+    return false
+  } else return true
+}
+
 /**
  * Clicks the 'Add Translation' button
  */
 function addTranslation () {
-  document.getElementsByClassName('key-add-suggestion-header').item(0).click()
+  if (!inputWrapCollapsed()) { // Don't toggle 'add' and 'cancel'
+    document.getElementsByClassName('key-add-suggestion-header').item(0).click()
+  } else { // Give focus back to input box (don't lose what was typed)
+    document.getElementsByClassName('input form-control tr-form-control key-add-suggestion-field').item(0).focus()
+  }
 }
 
 /**
- * Clicks the edit icon
+ * Clicks the edit icon.
+ * Useful when a suggestion exists but no translation is applied.
  */
 function editTranslation () {
-  document.getElementsByClassName('ibtn key-suggestion-edit').item(0).click()
+  // Don't click if already typing something
+  if (!inputWrapCollapsed()) {
+    document.getElementsByClassName('ibtn key-suggestion-edit').item(0).click()
+  } else { // Give focus back to input box (don't lose what was typed)
+    document.getElementsByClassName('input form-control tr-form-control key-add-suggestion-field').item(0).focus()
+  }
 }
 
+/**
+ * Clicks the cancel button
+ */
+function cancelTranslation () {
+  // Prevent possible glitch: Don't cancel if wrapper is collapsed, but focus is in a 'form-control' element
+  if (!inputWrapCollapsed) {
+    // Don't allow input-forms to keep the focus after cancel or else shortcuts won't work
+    document.getElementsByClassName('btn btn-default form-cancel-btn').item(0).focus()
```

Comment:
Curious, why are you focusing before clicking?

Reply:
should I focus after the click then?

Reply:
The keyboard shortcuts that had been prevented inside the input-form, don't work again after "clicking" CANCEL

That's because when the cancel button was simply "clicked", the input-wrapper collapses, the focus still remains inside the input-form. So whatever keys you pressed after 'cancel' they would go in the input form, and not trigger any of the other shortcuts.

If the click came from a finger/mouse press, then the focus would change to outside the input-form.